doc/_scripts/software_maturity_scanner.py

```diff
@@ -0,0 +1,554 @@
+
+
+def find_maturity_level(rule: list, experimental_symbols: Set[str]) -> str:
+    """Find if any of the symbols used in a rule are experimental or not.
+
+    Returns:
+        'Experimental' | 'Supported'
+    """
+
+    def flatten(items):
+        """Flatten the rule list to a set of unique symbols."""
+        out = set()
+        for item in items:
+            if isinstance(item, list):
+                out.update(flatten(item))
+            else:
+                out.add(item)
+        return out
+
+    symbols = flatten(rule)
+    for symbol in symbols:
+        if symbol in experimental_symbols:
+            return "Experimental"
+    return "Supported"
+
+
+def generate_tables(args) -> None:
+    """Entry point of software maturity level scanning.
+
+    Args:
+        args: CLI arguments.
+    """
+
+    # Directory sanity check
+    sample_dir = NRF_SAMPLE_DIR
+    if not args.build_output.exists():
+        sys.exit("Invalid output directory")
+    elif (
+        not (sample_dir / "samples").exists()
+        or not (sample_dir / "applications").exists()
+    ):
+        sys.exit("samples/ or applications/ cannot be found in " + str(sample_dir))
+
+    # Read input file
+    try:
+        with open(args.input_file) as f:
+            input_doc = yaml.safe_load(f)
+    except FileNotFoundError as fnfe:
+        sys.exit(f"Could not find input file '{fnfe.filename}'")
+    except yaml.YAMLError as yamlerr:
+        sys.exit("Invalid or corrupt input file: " + str(yamlerr))
+
+    # Create a dictionary of all unique features and their respective rule
+    all_features = {}
+    feature_categories = input_doc["features"]
+    for cat, features in feature_categories.items():
+        for feature, rule in features.items():
+            key = f"{cat}_{feature}"
+            all_features[key] = parse_rule(rule)
+
+    # Find all sample.yaml files
+    all_samples = list(sample_dir.glob("samples/**/sample.yaml"))
+    all_samples += list(sample_dir.glob("applications/**/sample.yaml"))
+
+    # Find all SoCs from integration_platforms in sample.yaml
+    all_socs = find_all_socs(all_samples)
+
+    # Check which symbols are experimental by looking through build.log files
+    build_logs = list(args.build_output.glob("**/build.log"))
+    exp_soc_sets = find_experimental_symbols(build_logs)
+
+    # Each feature is associated with a dictionary mapping each supported SoC
+    # to its state (Experimental/Supported)
+    soc_sets = {}
+    top_table_info = input_doc["top_table"]
+    for tech in top_table_info:
+        soc_sets[tech] = set()
+
+    for feature in all_features:
+        soc_sets[feature] = {}
+
+    # Look for the desired symbol combinations in all .config files
+    all_conf_files = args.build_output.glob("**/zephyr/.config")
+    for conf_file in all_conf_files:
+        kconf = kconfiglib.Kconfig(filename=conf_file)
+        soc = kconf.variables["CONFIG_SOC"].value
+        soc = soc.strip('"').split("_")[0].lower()
```

Comment:
`soc.strip('"').split("_")[0].lower()` is duplicated in this file, so you can extract it to separate method

Reply:
Added a `soc_from_kconfig` function which encompasses this and the previous line